### DISCUSSION TITLE
Can we hide or hash the model details?

### DISCUSSION THREAD

**Discussion by benyx6** [livewire/livewire#2627] [3. Feature ideas]
Hello guys, thanks for this great job 👏 

Please I realised that someone viewing the webpage source code can reverse engineer  my entire database!

**"dataMeta":{"models":{"permission":{"class":"Spatie\\Permission\\Models\\Permission","id":11,"relations":[],"connection":"mysql"}**

We give away the model details
we give away the model relationships 
worse of it all we tell what database is at the backend.

Can anything be done about this ?

Many thanks to you for your great efforts 👏  I Love ❤️ Livewire

**Comment by sfilzek** [comment]
Heya,

I agree completely with encrypting the model data. I’ve seen a bunch of discussions with people saying that there is nothing in the model data that could be used to exploit your app, and they may be correct, but it still leaks some internals which I’m not comfortable with. Even leaking the id can lead to people being able to gain some information about the size/growth of your app (use uuids!).

I understand the hesitation with the overhead of adding an encrypt/decrypt layer to the dataMeta section, but it would be great if this were a config option which could be enabled. The browser side of Livewire doesn’t really need to know the contents of the models info, afaik it is only used when hydrating the model on the server side, so apart from the cypher overhead there shouldn’t be an issue with this, correct?

Sab.

**Reply by benyx6** [reply]
Hello sfilzek, 
Thanks so much for your reply.
this is why I did not open it up as a security issue. And you are perfectly correct, imagine someone monitoring your database table growth rate by inspecting your IDs. 

Your suggestion is very brilliant, a config flag could be added to the Livewire config to enable or disable the encryption feature. So based on one's demands, one can decide to enable or disable it.

Yes there should not be any issues with adding this feature at all. 

Thanks!
Ben

**Comment by jelib3an** [comment]
Hi! Not sure if this helps, but we use this to hide things from the frontend. https://laravel.com/docs/8.x/eloquent-serialization#hiding-attributes-from-json

Then we make them visible only on components that need them.

**Reply by benyx6** [reply]
Hello Jelb3an, 
Thanks for you reply, I think Livewire does not respect those declarations in the model. 
Worse of it is, if you use third party libraries like Livewire-datatables, they just use the RAW DB, and so on. so they hardly use the model, rather they use the builder directly. This is good though because of performance. But then they end up exposing your table schema.

thanks
Ben

**Reply by francoism90** [reply]
@benyx6 Sorry for the bump, but it actually hides the attributes.

**Comment by johanmolen** [comment]
@benyx6 did you already find some kind of solution for this :)?

**Reply by benyx6** [reply]
@johanmolen I just placed an object between the model and the livewire components and manually populated the object with only the fields I needed while loosing some goodness of livewire. But for data tables like  https://github.com/mediconesystems/livewire-datatables, I am still not able to circumvent it because I have to rewrite the library which I am not ready for.

**Reply by johanmolen** [reply]
@benyx6 that also hides the whole dataMeta with model data and id's? Im using route model binding so for this approach is a little bit harder for me then.

**Reply by benyx6** [reply]
Hello, Johan @johanmolen , Sorry for the late response. Yes it does hide them except you decide to add it as part of your global variable that you want to expose. It is really harder because you have to manually do your thing.

It looks like they do not want to do anything about it.

**Comment by RenoLooijmans** [comment]
I'm also not very amused that Livewire exposes the internals of the app and came accross this discussion... Is there any progress made on this possible feature?

**Reply by RenoLooijmans** [reply]
I appear to get it working by using this piece of code. No internals are exposed at all. You need to check modelUuid though to be sure it is not tampered with during requests. This can be done with AuthorizesRequest trait for example.

```
<?php

namespace App\Http\Livewire;

use Livewire\Component;
use Models\Model;

class TestComponent extends Component
{
    public int $modelUuid;

    public function test()
    {
        // $this->model is now accessible in the component and in the livewire blade view.
    }

    public function getModelProperty()
    {
        return Model::where($this->modelUuid)->firstOrFail();
    }

    ...
}
```

`@livewire('path.test-component', ['modelUuid' => $this->model->uuid])`

I've not encountered any downsides yet.

Based on this discussion (which ironically proposed exposing the internals): https://github.com/livewire/livewire/issues/488

**Reply by francoism90** [reply]
It's not very flexible, see model binding. I also think you full model is still accessible, since you're using a public property.

**Comment by francoism90** [comment]
Why aren't the properties encrypted? This is not something I like.

**Comment by nicolaibaaring** [comment]
@calebporzio Is there a chance that this issue of model hydration can be solved by making changes to use the route key name instead of the model primary key?

Currently the ModelIdentifier uses `getQueueableId` to find the model. This means that the model id will always be exposed in the payloads. When setting the model's primary key to a different value, the whole database structure needs to be using this primary key for relationships to work - even when the only goal is to hide the model id from the public eyes, exactly as with a route model binding. If model hydration on subsequent requests used the defined route key instead, then the underlying application and database structure could be completely hidden.
Let me know, if I just missed something. :-)

Albeit, @RenoLooijmans solution works, it removes the possibility to pass models through bindings to the nested components. Without that being possible, I hardly find it a solution. My only current solution is to pass Wireable ViewModels to nested components, but it removes a lot of the ease of using Livewire.

**Comment by dimitri-koenig** [comment]
Hi guys

I've stumbled upon this issue too, with some additional performance issues. Until there is a livewire-core solution, hopefully this solution maybe helpful to some: https://issue.dev/livewire-mount-performance/

**Reply by francoism90** [reply]
Thanks for sharing. :)

However this kinda defeats the [model](https://laravel-livewire.com/docs/2.x/properties#binding-models)/[route](https://laravel-livewire.com/docs/2.x/rendering-components#route-model-binding) binding that is possible with Livewire.

It would be so much easier if Livewire would simple encrypt/decrypt the model string, model id('s), etc. Or just not make every public method accessible to the client.

**Reply by francoism90** [reply]
This seems to work:
```php
// App/Traits/WithUser

trait WithUser
{
    protected ?User $user = null;

    public function getUserProperty(): ?User
    {
        return $this->user;
    }
}
```

```php
// SomeController with route binding

class UserController
{
     use WithUser;

     public function mount(User $user)
     {
          $this->user = $user;
     }
}
```

In your blade you can call the computed property instead.

This way you cannot get a log of the model. However it's still possible to get properties, so I would recommend to play with model `visible`/`hidden`, e.g. hide `id` by default and use UUIDs instead.

**Comment by michael-rubel** [comment]
How about using this package?
https://github.com/archtechx/livewire-access

UPD: don't use the package, make the properties `protected` and access it in templates using `$this->property`.

**Reply by dimitri-koenig** [reply]
This is imho obsolete since you can access protected methods/properties within your template. This was not possible in earlier days, thus this package.

**Reply by michael-rubel** [reply]
@dimitri-koenig you can't properly access protected objects by default.

**Reply by dimitri-koenig** [reply]
Sure you can, see this PR: https://github.com/livewire/livewire/pull/3640

But just to make sure we are talking about the same thing: you can access protected properties/methods from within templates, but not from client side javascript.

**Reply by michael-rubel** [reply]
@dimitri-koenig wow, I didn't know that. Good to mention it requires the usage of `$this` in templates instead of just variable.

**Reply by michael-rubel** [reply]
@dimitri-koenig updated the best practices, thanks!
https://github.com/michael-rubel/livewire-best-practices/commit/8401541388119f5a5863647c9c9cd3f0c58d6e39

**Reply by francoism90** [reply]
@michael-rubel Thanks for updating the best practices! It makes me wonder how many Livewire-apps in the wild are 'insecure' by default.

**Comment by francoism90** [comment]
Could you guys please checkout https://github.com/livewire/livewire/pull/5742

This seems to encrypt the model data, it's still POC.
